### ISSUE TITLE
TFRecordDataset is loaded into memory entirely, yielding out-of-memory

### ISSUE THREAD

**Issue by jonyvp** [tensorflow/tensorflow#37630]
**System information** 
System information

Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): Yes
OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04): complex_model_m_gpu tier on the nightly docker image of tf
Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if
the issue happens on mobile device: /
TensorFlow installed from (source or
binary): - TensorFlow version (use command below): Nightly
Python version: - 3.6.9
version (if compiling from source): /
GCC/Compiler version (if compiling from
source): /
CUDA/cuDNN version: - GPU model and memory: Using K80 GPU's

**Describe the current behavior**
I previously loaded my images as npz bytestrings into a TFRecord, together with the label of the image. I would then convert the image and label to the expected format using TFRecordDataset mappings onto `tf.py_function()`s for training . 

I now try to do the formatting beforehand. This yields TFRecords that are much larger. I therefore use the `GZIP` compression. At runtime though, it seems the entire TFRecord is loaded into memory, making the training crash after some time. I even split the TFRecords into smaller partitions, but this makes no difference. How should this be tackled and/or is this normal behavior?

![image](https://user-images.githubusercontent.com/39876179/76743276-ab887180-6772-11ea-9017-5de8f0683a36.png)

**Comment by jonyvp** [comment]
Just did a rerun with the previous version of my TFRecords and code: Same issue. So it is not related to the compression. Could this be an issue with a new implementation of the `TFRecordDataset` in the nightly version?

**Comment by Saduf2019** [comment]
@jonyvp 
please share simple stand alone code for us to replicate this issue in our local

**Comment by jonyvp** [comment]
The issue doesn't seem related to any compression, but to the `shuffle()` being called on the TFRecordDataset. As the data in the TFRecord is now much larger than before (as all pre_processing operations are performed on it), the larger objects are loaded in the shuffle `buffer_size` and thus taking in much more space. Furthermore, during training, it seems more RAM is being used throughout the epochs. 

Is there any advice on the shuffle and buffer_size? A smaller buffer_size yields a non-perfect shuffle.

**Comment by jvishnuvardhan** [comment]
@jonyvp Can you please share a standalone code to reproduce the issue? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by jvishnuvardhan** [comment]
Closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
